File: sync-service/system_monitor.py
"""System metrics (CPU / RAM / GPU) for the monitor gadget.

CPU and memory come from psutil. GPU usage is read from nvidia-smi when it is
available on PATH; on machines without an NVIDIA GPU (or without the driver
tools) the ``gpu`` block reports ``available: False`` and the gadget hides its
GPU chart.
"""

from __future__ import annotations

import shutil
import subprocess
import sys
import time

import psutil

# Prime the CPU counter so the first real reading reflects usage since startup
# rather than returning 0.0 for the whole interval.
psutil.cpu_percent(interval=None)

_NVIDIA_SMI = shutil.which("nvidia-smi")
_NO_WINDOW = 0x08000000 if sys.platform == "win32" else 0

# nvidia-smi spawns a process per call; cache briefly so rapid polls (or several
# clients) do not pile up subprocesses.
_GPU_CACHE_TTL = 1.0
_gpu_cache: dict | None = None
_gpu_cache_at = 0.0


def _query_gpu() -> dict:
    if not _NVIDIA_SMI:
        return {"available": False}

    try:
        result = subprocess.run(
            [
                _NVIDIA_SMI,
                "--query-gpu=utilization.gpu,utilization.memory,memory.used,memory.total,name",
                "--format=csv,noheader,nounits",
            ],
            capture_output=True,
            text=True,
            timeout=4,
            creationflags=_NO_WINDOW,
        )
        line = result.stdout.strip().splitlines()[0]
        util, _mem_util, mem_used, mem_total, name = (part.strip() for part in line.split(","))
        mem_used_mb = float(mem_used)
        mem_total_mb = float(mem_total)
        return {
            "available": True,
            "percent": float(util),
            "memPercent": round(mem_used_mb / mem_total_mb * 100, 1) if mem_total_mb else 0.0,
            "memUsedMb": mem_used_mb,
            "memTotalMb": mem_total_mb,
            "name": name,
        }
    except Exception:
        return {"available": False}


def _get_gpu() -> dict:
    global _gpu_cache, _gpu_cache_at
    now = time.monotonic()
    if _gpu_cache is not None and (now - _gpu_cache_at) < _GPU_CACHE_TTL:
        return _gpu_cache
    _gpu_cache = _query_gpu()
    _gpu_cache_at = now
    return _gpu_cache


def get_system_metrics() -> dict:
    memory = psutil.virtual_memory()
    return {
        "cpu": {
            "percent": psutil.cpu_percent(interval=None),
            "cores": psutil.cpu_count(logical=True),
        },
        "memory": {
            "percent": memory.percent,
            "usedGb": round(memory.used / 1e9, 2),
            "totalGb": round(memory.total / 1e9, 2),
        },
        "gpu": _get_gpu(),
    }

File: sync-service-rs/src/system_monitor.rs
use std::process::Command;
use std::sync::Mutex;
use std::time::{Duration, Instant};

use serde_json::{json, Value};
use sysinfo::System;

static GPU_CACHE: Mutex<Option<(Instant, Value)>> = Mutex::new(None);

pub fn get_system_metrics() -> Value {
    let mut sys = System::new();
    sys.refresh_cpu_usage();
    sys.refresh_memory();

    // Brief yield so cpu usage is meaningful on first call patterns
    std::thread::sleep(Duration::from_millis(50));
    sys.refresh_cpu_usage();

    let cpu_percent = sys.global_cpu_usage() as f64;
    let cores = sys.cpus().len();
    let total = sys.total_memory() as f64;
    let used = sys.used_memory() as f64;
    let mem_percent = if total > 0.0 {
        (used / total) * 100.0
    } else {
        0.0
    };

    json!({
        "cpu": {
            "percent": (cpu_percent * 10.0).round() / 10.0,
            "cores": cores,
        },
        "memory": {
            "percent": (mem_percent * 10.0).round() / 10.0,
            "usedGb": (used / 1e9 * 100.0).round() / 100.0,
            "totalGb": (total / 1e9 * 100.0).round() / 100.0,
        },
        "gpu": get_gpu(),
    })
}

fn get_gpu() -> Value {
    let mut cache = GPU_CACHE.lock().unwrap();
    if let Some((at, value)) = cache.as_ref() {
        if at.elapsed() < Duration::from_secs(1) {
            return value.clone();
        }
    }
    let value = query_gpu();
    *cache = Some((Instant::now(), value.clone()));
    value
}

fn query_gpu() -> Value {
    let mut command = Command::new("nvidia-smi");
    command.args([
        "--query-gpu=utilization.gpu,utilization.memory,memory.used,memory.total,name",
        "--format=csv,noheader,nounits",
    ]);
    crate::process_win::hide_console(&mut command);
    let output = command.output();

    let Ok(output) = output else {
        return json!({ "available": false });
    };
    if !output.status.success() {
        return json!({ "available": false });
    }
    let stdout = String::from_utf8_lossy(&output.stdout);
    let Some(line) = stdout.lines().next() else {
        return json!({ "available": false });
    };
    let parts: Vec<&str> = line.split(',').map(|s| s.trim()).collect();
    if parts.len() < 5 {
        return json!({ "available": false });
    }
    let util: f64 = parts[0].parse().unwrap_or(0.0);
    let mem_used: f64 = parts[2].parse().unwrap_or(0.0);
    let mem_total: f64 = parts[3].parse().unwrap_or(0.0);
    let name = parts[4].to_string();
    let mem_percent = if mem_total > 0.0 {
        (mem_used / mem_total * 1000.0).round() / 10.0
    } else {
        0.0
    };
    json!({
        "available": true,
        "percent": util,
        "memPercent": mem_percent,
        "memUsedMb": mem_used,
        "memTotalMb": mem_total,
        "name": name,
    })
}
